packages/core/crates/core-crypto/src/prp.rs

```diff
@@ -0,0 +1,288 @@
+use crate::derivation::generate_key_iv;
+use crate::errors::Result;
+use crate::errors::CryptoError::InvalidInputValue;
+
+use crate::primitives::{calculate_mac, SimpleStreamCipher};
+
+// Module-specific constants
+const PRP_INTERMEDIATE_KEY_LENGTH: usize = 32;
+const PRP_INTERMEDIATE_IV_LENGTH: usize = 16;
+const PRP_KEY_LENGTH: usize = 4 * PRP_INTERMEDIATE_KEY_LENGTH;
+const PRP_IV_LENGTH: usize = 4 * PRP_INTERMEDIATE_IV_LENGTH;
+const HASH_KEY_PRP: &str = "HASH_KEY_PRP";
+
+// The minimum input length must be at least size of the key, which is XORed with plaintext/ciphertext
+pub const PRP_MIN_LENGTH: usize = PRP_INTERMEDIATE_KEY_LENGTH;
+
+/// Parameters for the Pseudo-Random Permutation (PRP) function
+/// This consists of IV and the raw secret key for use by the underlying cryptographic transformation.
+#[cfg_attr(feature = "wasm", wasm_bindgen::prelude::wasm_bindgen)]
+pub struct PRPParameters {
+    key: [u8; PRP_KEY_LENGTH],
+    iv: [u8; PRP_IV_LENGTH]
+}
+
+impl Default for PRPParameters {
+    fn default() -> Self {
+        Self {
+            key: [0u8; PRP_KEY_LENGTH],
+            iv: [0u8; PRP_IV_LENGTH]
+        }
+    }
+}
+
+#[cfg_attr(feature = "wasm", wasm_bindgen::prelude::wasm_bindgen)]
+impl PRPParameters {
+    /// Creates new parameters for the PRP by expanding the given
+    /// keying material into the secret key and IV for the underlying cryptographic transformation.
+    #[cfg_attr(feature = "wasm", wasm_bindgen::prelude::wasm_bindgen(constructor))]
+    pub fn new(secret: &[u8]) -> Self {
+        let mut ret = PRPParameters::default();
+        generate_key_iv(secret, HASH_KEY_PRP.as_bytes(), &mut ret.key, &mut ret.iv, false)
+            .expect("invalid secret given");
+        ret
+    }
+}
+
+/// Implementation of Pseudo-Random Permutation (PRP).
+/// Currently based on the Lioness wide-block cipher.
+#[cfg_attr(feature = "wasm", wasm_bindgen::prelude::wasm_bindgen)]
+pub struct PRP {
+    keys: [Vec<u8>; 4],
```

Comment:
Shouldn't this rather be a Box? Could be misused if length of vector is arbitrary

Reply:
Certainly can make it a `Box`